src/py/reactpy/reactpy/core/hooks.py

```diff
@@ -61,23 +75,62 @@ def use_state(initial_value: _Type | Callable[[], _Type]) -> State[_Type]:
     Returns:
         A tuple containing the current state and a function to update it.
     """
-    current_state = _use_const(lambda: _CurrentState(initial_value))
+    if server_only:
+        key = None
+    else:
+        hook = get_current_hook()
+        caller_info = get_caller_info()
+        key = get_state_key(caller_info)
+        if hook.reconnecting.current:
+            try:
+                initial_value = hook.client_state[key]
+            except KeyError as err:
+                raise StateRecoveryFailureError(
+                    f"Missing expected key {key} on client"
+                ) from err
+    current_state = _use_const(lambda: _CurrentState(key, initial_value))
     return State(current_state.value, current_state.dispatch)
 
 
+def get_caller_info():
+    # Get the current stack frame and then the frame above it
+    caller_frame = sys._getframe(2)
+    for i in range(50):
+        render_frame = sys._getframe(4 + i)
+        patch_path = render_frame.f_locals.get("patch_path_for_state")
+        if patch_path is not None:
+            break
+    # Extract the relevant information: file path and line number and hash it
+    return f"{caller_frame.f_code.co_filename} {caller_frame.f_lineno} {patch_path}"
```

Comment:
Ugly hack - to preserve the interface, it walks up the stack frames to find `patch_path_for_state` in the locals. I would suggest refactoring it

Reply:
This returns the unique identifier of each state / ref/ hook call based on what called it. This is eventually hashed and then the hash is truncated to create the key. The purpose of truncate the hash is to make it more challenging to reproduce this information.